### PR TITLE
Resolves #789 - generate screenshot at publish-time and upload to server

### LINKED ISSUE #789: Automatically generate previews for graphs
dimensions are 440x330

### PR DESCRIPTION
**Pull request by kkananen**

### CODE REVIEW

File: lib/previewImageProcessor.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+	if (image) {
+		var imgString = image.replace(/^data:image\/\w+;base64,/, "")
+		var imgData = new Buffer(imgString, 'base64')
+
+		var that = this
+
+		gm(imgData, "img.png").flip().toBuffer('PNG', function(err, buffer) {
+			if (err) {
+				dfd.resolve(that.defaultImage)
```
Comment (kschzt) [inline lib/previewImageProcessor.js]: should at least log the error here, could be very serious if this failed

File: models/graph.js [hunk 2]
```diff
@@ -18,6 +18,7 @@ var graphSchema = new mongoose.Schema({
 		match: alphanumeric,
 		index: true
 	}],
+	image: { type: String },
```
Comment (kschzt) [inline models/graph.js]: call this `imageUrl`?

Comment (kschzt) [inline models/graph.js]: actually, `previewUrl` rather?

File: controllers/graphController.js [hunk 3]
```diff
@@ -270,14 +272,24 @@ GraphController.prototype.save = function(req, res, next) {
 				.json({message: 'Sorry, permission denied'});
 		}
 
-		return that._fs.writeString(gridFsPath, req.body.graph)
+		return that._fs.writeString(gridFsGraphPath, req.body.graph)
+		.then(function() {
+			var previewImageProcessor = new PreviewImageProcessor()
+
+			return previewImageProcessor.process(req.body.previewImage)
+			.then(function(processedImage) {
```
Comment (kschzt) [inline controllers/graphController.js]: please write a test case for this path

File: controllers/graphController.js [hunk 4]
```diff
@@ -270,14 +272,24 @@ GraphController.prototype.save = function(req, res, next) {
...
-			var url = that._fs.url(gridFsPath);
+			var url = that._fs.url(gridFsGraphPath);
+			var previewUrl = that._fs.url(gridFsPreviewPath)
 
 			var model = {
 				path: path,
 				tags: tags,
-				url: url
+				url: url,
+				image: previewUrl
```
Comment (kschzt) [inline controllers/graphController.js]: and a test case for this path

File: lib/previewImageProcessor.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+PreviewImageProcessor.prototype.process = function(path, image) {
+	var dfd = when.defer()
+
+	if (image) {
+		var imgString = image.replace(/^data:image\/\w+;base64,/, "")
+		var imgData = new Buffer(imgString, 'base64')
+
+		var that = this
+
+		gm(imgData, "img.png").flip().toBuffer('PNG', function(err, buffer) {
```
Comment (gmarinov) [inline lib/previewImageProcessor.js]: how does gm() handle garbage? 
shouldn't trust incoming data..

**Comment by kschzt** [comment]
lgtm

**Pull request merged**